zips/zip-0216.rst

```diff
@@ -107,74 +114,167 @@ In transactions [#protocol-txnencoding]_:
+in a Sapling payment address, or $\mathsf{pk\star_d}$ in a decrypted $\mathsf{C^{out}}$,
+to encode the zero point. The history is explained in a note added in version 2023.4.0 of
+the protocol specification [#protocol-2023.4.0]_:
+
+  A previous version of this specification did not have the requirement for the decoded point
+  $\mathsf{pk_d}$ of a Sapling note to be in the set of prime‑order points $\mathbb{J}^{(r)*}$
+  (i.e. "if ... $\mathsf{pk_d} \not\in \mathbb{J}^{(r)*}$, return $\bot\!$").
+  That did not match the implementation in `zcashd`. In fact the history is a little more
+  complicated. The current specification matches the implementation in `librustzcash` as of
+  [#librustzcash-pr109]_, which has been used in `zcashd` since v2.1.2. However, there was
+  another implementation of Sapling note decryption used in `zcashd` for consensus checks,
+  specifically the check that a shielded coinbase output decrypts successfully with the
+  zero $\mathsf{ovk}$. This was corrected to enforce the same restriction on the decrypted
+  $\mathsf{pk_d}$ in `zcashd` v5.5.0, originally set to activate in a soft fork at block
+  height 2121200 on both Mainnet and Testnet [#zcashd-pr6459]_. (On Testnet this height was
+  in the past as of the `zcashd` v5.5.0 release, and so the change would have been immediately
+  enforced on upgrade.) Since the soft fork was observed to be retrospectively valid after
+  that height, the implementation was simplified in [#zcashd-pr6725]_ to use the
+  `librustzcash` implementation in all cases, which reflects the specification above.
+  `zebra` always used the `librustzcash` implementation.
+
+The protocol specification was further updated in version 2025.6.0 [#protocol-2025.6.0]_,
+to tighten the type of $\mathsf{ivk}$ in Sapling to
+$\{ 1\,..\, 2^{\ell^{\mathsf{Sapling}}_{\mathsf{ivk}}}\!-1 \}$ and the type of $\mathsf{pk_d}$
+in both Sapling and Orchard to $\mathsf{KA^{protocol}.PublicPrimeOrder}$, in order to make
+the exclusion of the zero point more obvious [#zips-issue664]_. This also has the effect
+that a Sapling incoming viewing key [#protocol-saplinginviewingkeyencoding]_ or a Sapling
+IVK Encoding in a Unified Incoming Viewing Key [#zip-0316]_ that encodes the zero
+$\mathsf{ivk}$ MUST be considered invalid when imported. Another note was also added
+explicitly covering the encoding of $\mathsf{pk_d}$ in Sapling payment addresses
+[#protocol-saplingpaymentaddrencoding]_:
+
+  The restriction on $\mathsf{pk_d}$ reflects its current type
+  $\mathsf{KA^{Sapling}.PublicPrimeOrder} = \mathbb{J}^{(r)*}$. In versions of this
+  specification prior to 2025.6.0, $\mathsf{pk_d}$ had type
+  $\mathsf{KA^{Sapling}.PublicPrimeSubgroup} = \mathbb{J}^{(r)}$, i.e. including
+  $\mathcal{O}_{\mathbb{J}}$. Implementations of consumers for this encoding may need
+  to be updated to exclude $\mathcal{O}_{\mathbb{J}}$, and should be checked for
+  consistency with the current version of [ZIP-216].
```

Comment:
I'm not sure what librustzcash does when reading an $\mathsf{ivk}$; we should check this. If it always reconstructs the $\mathsf{pk_d}$ soon afterward then a zero $\mathsf{ivk}$ should be caught by that.